lib/livebook/web_socket.ex

```diff
@@ -0,0 +1,96 @@
+defmodule Livebook.WebSocket do
+  @moduledoc false
+
+  alias Livebook.WebSocket.{
+    Client,
+    Request,
+    SessionRequest
+  }
+
+  defmodule Connection do
+    defstruct [:conn, :websocket, :ref]
+
+    @type t :: %__MODULE__{
+            conn: Client.conn(),
+            websocket: Client.websocket(),
+            ref: Client.ref()
+          }
+  end
+
+  @app_version Mix.Project.config()[:version]
+
+  @typep header :: {String.t(), String.t()}
+  @typep headers :: list(header())
+
+  @doc """
+  Connects with the WebSocket server for given URL and headers.
+  """
+  @spec connect(String.t(), headers()) ::
+          {:ok, Connection.t(), :connected | {atom(), struct()}}
+          | {:error, Connection.t(), String.t() | Livebook.WebSocket.Error.t()}
+  def connect(url, headers \\ []) do
+    with {:ok, conn, ref} <- Client.connect(url, headers) do
+      conn
+      |> Client.receive(ref)
+      |> handle_receive(ref)
+    end
+  end
+
+  @doc """
+  Disconnects the given WebSocket client.
+  """
+  @spec disconnect(Connection.t()) :: :ok
+  def disconnect(%Connection{} = connection) do
+    Client.disconnect(connection.conn, connection.websocket, connection.ref)
+  end
+
+  @dialyzer {:nowarn_function, send_session: 1}
+
+  @doc """
+  Sends a session request to the given server.
+  """
+  @spec send_session(Connection.t()) :: Client.send_fun()
+  def send_session(%Connection{} = connection) do
+    session_request = SessionRequest.new!(app_version: @app_version)
+    message = Request.new!(type: {:session, session_request})
```

Comment:
It would be nice if we could generalize this function so we only pass the Request object but I see we need to use tagged tuples. Would it be possible to recover the tuple tag using some reflection API?

Reply:
We could generalize it as: `send_request(connection, :session, SessionRequest.new!(...))`

But it would be nice if we could skip the second argument. We don't need to handle this in this PR.